test/gtest/ucp/test_ucp_peer_failure.cc

```diff
@@ -570,3 +621,163 @@ UCS_TEST_P(test_ucp_peer_failure_keepalive, kill_receiver,
 }
 
 UCP_INSTANTIATE_TEST_CASE(test_ucp_peer_failure_keepalive)
+
+
+struct test_ucp_peer_failure_rndv_abort : public test_ucp_peer_failure {
+    void init()
+    {
+        self = this;
+
+        if (has_any_transport({"ud_v", "ud_x", "shm,ib", "ugni", "self", "tcp"})) {
+            UCS_TEST_SKIP_R("No rndv/put/zcopy support");
+        }
+
+        if (!m_ucp_config->ctx.proto_enable) {
+            UCS_TEST_SKIP_R("Test is designed to work with protov2 only");
+        }
+
+        test_ucp_peer_failure::init();
+
+        receiver().connect(&sender(), get_ep_params());
+        sender().connect(&receiver(),  get_ep_params());
+        set_am_handler(receiver());
+
+        m_sbuf.resize(UCS_KBYTE);
+        m_rbuf.resize(UCS_KBYTE);
+    }
+
+
+    static void get_test_variants(std::vector<ucp_test_variant> &variants)
+    {
+        add_variant_with_value(variants, UCP_FEATURE_AM, TEST_AM, "am");
+        add_variant_with_value(variants, UCP_FEATURE_TAG, TEST_TAG, "tag");
+    }
+
+    void protos_replace_progress()
+    {
+        for (int proto_id = 0; proto_id < ucp_protocols_count(); ++proto_id) {
+            auto proto = const_cast<ucp_proto_t*>(ucp_protocols[proto_id]);
+            m_progress_storage.emplace(proto, progress_array{});
+
+            int stage = UCP_PROTO_STAGE_START;
+            for (; stage < UCP_PROTO_STAGE_LAST; ++stage) {
+                m_progress_storage[proto][stage] = proto->progress[stage];
+                proto->progress[stage]           = &progress_wrapper;
+            }
+        }
+    }
+
+    void protos_restore_progress()
+    {
+        for (int proto_id = 0; proto_id < ucp_protocols_count(); ++proto_id) {
+            auto proto = const_cast<ucp_proto_t*>(ucp_protocols[proto_id]);
+
+            int stage = UCP_PROTO_STAGE_START;
+            for (; stage < UCP_PROTO_STAGE_LAST; ++stage) {
+                proto->progress[stage] = m_progress_storage[proto][stage];
+            }
+        }
+    }
```

Comment:
can this be unified somehow? The code is almost identical

Reply:
Good point! Done.